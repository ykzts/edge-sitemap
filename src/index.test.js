import { expect, test } from '@jest/globals'
import { SitemapStream } from './index.js'

/**
 * @param {import('.').SitemapItem[]} sitemapItems
 * @param {WritableStream<import('.').SitemapItem>} writableStream
 */
async function writeSitemapItems(sitemapItems, writableStream) {
  const writer = writableStream.getWriter()

  await writer.ready

  await Promise.all(
    sitemapItems.map((sitemapItem) => writer.write(sitemapItem))
  )

  await writer.close()
}

/**
 * @param {ReadableStream<string>} readableStream
 */
async function readableStreamToString(readableStream) {
  const reader = readableStream.getReader()
  let result = ''

  /**
   * @returns {Promise<ReadableStreamReadResult<string> | void>}
   */
  async function pump() {
    const { done, value } = await reader.read()

    if (!done) {
      result += value

      return pump()
    }
  }

  await pump()

  return result
}

test('basic use case', () => {
  const { readable, writable } = new SitemapStream({
    baseURL: 'https://example.com/',
  })

  writeSitemapItems(
    [
      {
        loc: '/',
        priority: 1.0,
      },
      {
        loc: '/privacy',
        priority: 0.1,
      },
    ],
    writable
  )

  return expect(readableStreamToString(readable)).resolves.toMatchSnapshot()
})

test('pretty results', () => {
  const { readable, writable } = new SitemapStream({
    baseURL: 'https://example.com/',
    pretty: true,
  })

  writeSitemapItems(
    [
      {
        loc: '/',
        priority: 1.0,
      },
      {
        loc: '/privacy',
        priority: 0.1,
      },
    ],
    writable
  )

  return expect(readableStreamToString(readable)).resolves.toMatchSnapshot()
})
