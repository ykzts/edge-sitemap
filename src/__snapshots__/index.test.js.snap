// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic use case 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"><url><loc>https://example.com/</loc><priority>1.0</priority></url><url><loc>https://example.com/privacy</loc><priority>0.1</priority></url></urlset>"
`;

exports[`pretty results 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://example.com/</loc>
    <priority>1.0</priority>
  </url>
  <url>
    <loc>https://example.com/privacy</loc>
    <priority>0.1</priority>
  </url>
</urlset>
"
`;
